nest_elephant_tvb/translation/mpi_translator.py

```diff
@@ -3,192 +3,237 @@
+    ############ NEW Code: Receive/analyse/send
+    if comm.Get_rank() == 0: # Receiver from NEST, rank 0
+        _receive(comm_receiver, databuffer, logger_receive)
+    else: #  Science/analyse and sender to TVB, rank 1-x
+        _send(comm_sender, databuffer, logger_send, store, analyse)
+    ############ NEW Code end
     
-    ############ OLD Code to replace: with threads, current solution
-    initialisation =np.load(param['init']) # initialisation value
-    buffer=[initialisation] # buffer of the rate to send it
-        
-    # create the thread for receive and send data
-    th_receive = Thread(target=receive, args=(logger_receive,store,status_data,buffer, comm_receiver))
-    th_send = Thread(target=send, args=(logger_send,analyse,status_data,buffer, comm_sender))
-
-    # start the threads
-    # FAT END POINT
-    logger_master.info('Start thread')
-    th_receive.start()
-    th_send.start()
-    th_receive.join()
-    th_send.join()
-    logger_master.info('thread join')
-    ############# OLD Code end
+    ############ NEW Code: disconnect
+    # TODO: should this be done here?
+    logger_master.info('Disconnect communicators...')
+    comm_receiver.Disconnect()
+    comm_sender.Disconnect()
+    ############ NEW Code end
 
 
-def _shared_mem_buffer(buf_pkgs,sizes,comm):
+def _shared_mem_buffer(comm):
     '''
-    Create a shared memory block.
-    MPI One-sided-Communication.
-    NOTE: still copy-paste from kim master thesis code...
+    Create shared memory buffer. MPI One-sided-Communication.
+    :param comm: MPI intra communicator to create the buffer.
+    :return buffer: 1D shared memory buffer array 
+    
+    TODO: data from NEST not constant. 
+    Buffersize/max. expected number of events hardcoded. 3 doubles for each event.
+    
+    TODO: reshaping of the buffer content is done in the science part
+    -> 1D array to shape (x,3) where x is the number of events.
+    -> this is decided by the nest I/O part, i.e. nest sends out events in 1D array 
+    -> can/should this be more generic?
     '''
-    pkg_size = np.sum(sizes)
     datasize = MPI.DOUBLE.Get_size()
+    bufsize = 1000000 * 3 # NOTE: hardcoded (max.expected events per package from nest)
     if comm.Get_rank() == 0:
-        bufbytes = buf_pkgs * datasize*2 * pkg_size 
+        bufbytes = datasize * bufsize
     else: 
         bufbytes= 0
     # rank 0: create the shared block
     # rank 1-x: get a handle to it
     win = MPI.Win.Allocate_shared(bufbytes, datasize, comm=comm)
-    buf, datasize = win.Shared_query(0) 
+    buf, datasize = win.Shared_query(0)
     assert datasize == MPI.DOUBLE.Get_size()
     # create a numpy array (buffer) whose data points to the shared mem
-    return np.ndarray(buffer=buf, dtype='d', shape=(buf_pkgs*pkg_size,2))
+    return np.ndarray(buffer=buf, dtype='d', shape=(bufsize,))
 
 
 
-def receive(logger,store,status_data,buffer, comm):
+def _receive(comm_receiver, databuffer, logger):
```

Comment:
can you create a new file for this function?
I would like to have different file for Nest I/O, transformer and TVB I/O

Reply:
I suggest a single file solution for now: this makes the review easier. The next step would see the split in multiple files
Kim: could you add a TODO to move to separate file later?

Reply:
Yes, it is planned in the next refactor steps to move the functionalities and tasks to different files.

Added a TODO.